File: app/src/main/java/com/gmail/fattazzo/formula1world/fragments/current/drivers/detail/pages/ranking/RankingDriverFragment.java
package com.gmail.fattazzo.formula1world.fragments.current.drivers.detail.pages.ranking;

import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.view.View;
import android.widget.ProgressBar;

import com.github.mikephil.charting.animation.Easing;
import com.github.mikephil.charting.charts.LineChart;
import com.gmail.fattazzo.formula1world.R;
import com.gmail.fattazzo.formula1world.domain.F1Driver;
import com.gmail.fattazzo.formula1world.domain.F1DriverStandings;
import com.gmail.fattazzo.formula1world.domain.F1Result;
import com.gmail.fattazzo.formula1world.service.DataService;
import com.gmail.fattazzo.formula1world.utils.ImageUtils;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.Background;
import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.Click;
import org.androidannotations.annotations.EFragment;
import org.androidannotations.annotations.FragmentArg;
import org.androidannotations.annotations.UiThread;
import org.androidannotations.annotations.ViewById;
import org.apache.commons.collections4.CollectionUtils;

import java.util.ArrayList;
import java.util.List;

/**
 * @author fattazzo
 *         <p/>
 *         date: 21/04/17
 */
@EFragment(R.layout.fragment_driver_ranking)
public class RankingDriverFragment extends Fragment {

    @FragmentArg
    F1Driver driver;

    @Bean
    ImageUtils imageUtils;

    @Bean
    ChartManager chartManager;

    @Bean
    DataService dataService;

    @ViewById(R.id.ranking_driver_position_chart)
    LineChart positionChart;

    @ViewById(R.id.ranking_driver_points_chart)
    LineChart pointsChart;

    @ViewById
    FloatingActionButton refreshFab;

    @ViewById
    ProgressBar progressBar;

    private List<F1Result> raceResults, leaderRaceResults;

    public static RankingDriverFragment newInstance(F1Driver driver) {
        RankingDriverFragment rankingDriverFragment = new RankingDriverFragment_();
        Bundle args = new Bundle();
        args.putSerializable("driver", driver);
        rankingDriverFragment.setArguments(args);
        return rankingDriverFragment;
    }

    @AfterViews
    void init() {
        chartManager.configureChart(positionChart);
        chartManager.configureChart(pointsChart);

        loadData();
    }

    @Click
    void refreshFab() {
        dataService.clearDriverRaceResultsCache(driver);
        if (CollectionUtils.isNotEmpty(leaderRaceResults)) {
            dataService.clearDriverRaceResultsCache(leaderRaceResults.get(0).driver);
        }
        loadData();
    }

    @UiThread
    void startLoad() {
        if (progressBar != null) {
            progressBar.setVisibility(View.VISIBLE);
        }
    }

    @Background
    void loadData() {
        startLoad();

        raceResults = dataService.loadDriverRacesResult(driver);

        leaderRaceResults = new ArrayList<>();
        F1DriverStandings driverLeader = dataService.loadDriverLeader();
        if (driverLeader != null) {
            leaderRaceResults = dataService.loadDriverRacesResult(driverLeader.driver);
        }

        loadChartData();
    }

    @UiThread
    void loadChartData() {
        try {
            if (positionChart != null && pointsChart != null) {
                chartManager.loadPositionsChartData(positionChart, raceResults);
                chartManager.loadPointsChartData(pointsChart, raceResults, leaderRaceResults);
                startChatsAnimation();
            }
        } finally {
            if (progressBar != null) {
                progressBar.setVisibility(View.INVISIBLE);
            }
        }
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);

        if (isVisibleToUser && positionChart != null) {
            startChatsAnimation();
        }
    }

    private void startChatsAnimation() {
        positionChart.animateY(1000, Easing.EasingOption.EaseInOutSine);
        pointsChart.animateY(1000, Easing.EasingOption.EaseInOutSine);
    }

}

File: app/src/main/kotlin/com/gmail/fattazzo/formula1world/fragments/current/drivers/detail/pages/ranking/RankingDriverFragment.kt
package com.gmail.fattazzo.formula1world.fragments.current.drivers.detail.pages.ranking

import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v4.app.Fragment
import android.view.View
import android.widget.ProgressBar
import com.github.mikephil.charting.animation.Easing
import com.github.mikephil.charting.charts.LineChart
import com.gmail.fattazzo.formula1world.R
import com.gmail.fattazzo.formula1world.domain.F1Driver
import com.gmail.fattazzo.formula1world.domain.F1Result
import com.gmail.fattazzo.formula1world.service.DataService
import com.gmail.fattazzo.formula1world.utils.ImageUtils
import org.androidannotations.annotations.*
import org.apache.commons.collections4.CollectionUtils
import java.util.*

/**
 * @author fattazzo
 *
 *
 * date: 21/04/17
 */
@EFragment(R.layout.fragment_driver_ranking)
open class RankingDriverFragment : Fragment() {

    @FragmentArg
    lateinit internal var driver: F1Driver

    @Bean
    lateinit internal var imageUtils: ImageUtils

    @Bean
    lateinit internal var chartManager: ChartManager

    @Bean
    lateinit internal var dataService: DataService

    @ViewById(R.id.ranking_driver_position_chart)
    lateinit internal var positionChart: LineChart

    @ViewById(R.id.ranking_driver_points_chart)
    lateinit internal var pointsChart: LineChart

    @ViewById
    lateinit internal var refreshFab: FloatingActionButton

    @ViewById
    lateinit internal var progressBar: ProgressBar

    private var raceResults: List<F1Result>? = null
    private var leaderRaceResults: List<F1Result>? = null

    @AfterViews
    internal fun init() {
        chartManager!!.configureChart(positionChart!!)
        chartManager!!.configureChart(pointsChart!!)

        loadData()
    }

    @Click
    internal fun refreshFabClicked() {
        dataService!!.clearDriverRaceResultsCache(driver!!)
        if (CollectionUtils.isNotEmpty(leaderRaceResults)) {
            dataService!!.clearDriverRaceResultsCache(leaderRaceResults!![0].driver!!)
        }
        loadData()
    }

    @UiThread
    internal open fun startLoad() {
        if (progressBar != null) {
            progressBar!!.visibility = View.VISIBLE
        }
    }

    @Background
    internal open fun loadData() {
        startLoad()

        raceResults = dataService!!.loadDriverRacesResult(driver!!)

        leaderRaceResults = ArrayList()
        val driverLeader = dataService!!.loadDriverLeader()
        if (driverLeader != null) {
            leaderRaceResults = dataService!!.loadDriverRacesResult(driverLeader.driver!!)
        }

        loadChartData()
    }

    @UiThread(propagation = UiThread.Propagation.REUSE)
    internal open fun loadChartData() {
        try {
            if (positionChart != null && pointsChart != null) {
                if(raceResults!!.isNotEmpty()) {
                    chartManager!!.loadPositionsChartData(positionChart!!, raceResults!!)
                }
                if(leaderRaceResults!!.isNotEmpty()) {
                    chartManager!!.loadPointsChartData(pointsChart!!, raceResults!!, leaderRaceResults!!)
                }
                startChatsAnimation()
            }
        } finally {
            if (progressBar != null) {
                progressBar!!.visibility = View.INVISIBLE
            }
        }
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)

        if (isVisibleToUser && positionChart != null) {
            startChatsAnimation()
        }
    }

    private fun startChatsAnimation() {
        positionChart!!.animateY(1000, Easing.EasingOption.EaseInOutSine)
        pointsChart!!.animateY(1000, Easing.EasingOption.EaseInOutSine)
    }

    companion object {

        fun newInstance(driver: F1Driver): RankingDriverFragment {
            val rankingDriverFragment = RankingDriverFragment_()
            val args = Bundle()
            args.putSerializable("driver", driver)
            rankingDriverFragment.arguments = args
            return rankingDriverFragment
        }
    }

}
